GH4687: Trim leading hyphens from command-line argument keys in FrostingConfiguration

paulomorgado opened the pull request:
Trim leading hyphens from command-line argument keys in the
`FrostingConfiguration.cs` file to ensure consistent key
formatting when processing arguments. This change improves
compatibility and standardization in the `args` dictionary.

Addresses https://github.com/orgs/cake-build/discussions/4665

paulomorgado committed:
Trim leading hyphens from command-line argument keys in the
`FrostingConfiguration.cs` file to ensure consistent key
formatting when processing arguments. This change improves
compatibility and standardization in the `args` dictionary.

devlead commented on src/Cake.Frosting/Internal/FrostingConfiguration.cs:
```suggestion

```

paulomorgado replied:
Done!

devlead commented on src/Cake.Frosting/Internal/FrostingConfiguration.cs:
```suggestion
        public FrostingConfiguration(IEnumerable<FrostingConfigurationValue> values, IFileSystem fileSystem, ICakeEnvironment environment, ICakeArguments arguments)
```

paulomorgado replied:
Done!

devlead commented on src/Cake.Frosting/Internal/FrostingConfiguration.cs:
```diff
@@ -31,7 +31,7 @@ public FrostingConfiguration(IEnumerable<FrostingConfigurationValue> values, IFi
 
             var provider = new CakeConfigurationProvider(fileSystem, environment);
-            var args = remainingArguments.Parsed.ToDictionary(x => x.Key, x => x.FirstOrDefault() ?? string.Empty);
+            var args = remainingArguments.Parsed.ToDictionary(x => x.Key.TrimStart('-'), x => x.FirstOrDefault() ?? string.Empty);
```
```suggestion
            var args = arguments.GetArguments().ToDictionary(x => x.Key, x => x.Value?.FirstOrDefault() ?? string.Empty);
```

paulomorgado replied:
Done!

devlead commented on src/Cake.Frosting/Internal/FrostingConfiguration.cs:
```suggestion
            ArgumentNullException.ThrowIfNull(environment);

            ArgumentNullException.ThrowIfNull(arguments);
```

paulomorgado replied:
Done!

devlead reviewed (changes requested):
Like Cake.Tool and Cake.Sdk utilize `ICakeArguments` instead of Specre.Console.Cli `ICakeArguments` so we'vegot consistent behaviour.

```diff
@@ -8,7 +8,6 @@ using System.Linq;
 using Cake.Core;
 using Cake.Core.Configuration;
 using Cake.Core.IO;
-using Spectre.Console.Cli;
 
 namespace Cake.Frosting.Internal
 {
@@ -16,7 +15,7 @@ namespace Cake.Frosting.Internal
     {
         private readonly ICakeConfiguration _cakeConfiguration;
 
-        public FrostingConfiguration(IEnumerable<FrostingConfigurationValue> values, IFileSystem fileSystem, ICakeEnvironment environment, IRemainingArguments remainingArguments)
+        public FrostingConfiguration(IEnumerable<FrostingConfigurationValue> values, IFileSystem fileSystem, ICakeEnvironment environment, ICakeArguments arguments)
         {
             ArgumentNullException.ThrowIfNull(values);
 
@@ -24,6 +23,8 @@ namespace Cake.Frosting.Internal
 
             ArgumentNullException.ThrowIfNull(environment);
 
+            ArgumentNullException.ThrowIfNull(arguments);
+
             var baseConfiguration = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
             foreach (var value in values)
             {
@@ -31,7 +32,7 @@ namespace Cake.Frosting.Internal
             }
 
             var provider = new CakeConfigurationProvider(fileSystem, environment);
-            var args = remainingArguments.Parsed.ToDictionary(x => x.Key.TrimStart('-'), x => x.FirstOrDefault() ?? string.Empty);
+            var args = arguments.GetArguments().ToDictionary(x => x.Key, x => x.Value?.FirstOrDefault() ?? string.Empty);
 
             _cakeConfiguration = provider.CreateConfiguration(environment.WorkingDirectory, baseConfiguration, args);
         }

```

To test this you could add som configuration validation in Frosting integration test
https://github.com/cake-build/cake/tree/develop/tests/integration/Cake.Frosting/build

It's called here
https://github.com/cake-build/cake/blob/3531435a932dd112984f04798af694efa5326997/build.cake#L374-L384

devlead reviewed (changes requested):
To test this add som configuration validation in Frosting integration test project here:
https://github.com/cake-build/cake/tree/develop/tests/integration/Cake.Frosting/build

It's called here:
https://github.com/cake-build/cake/blob/3531435a932dd112984f04798af694efa5326997/build.cake#L374-L384

paulomorgado commented:
> To test this add som configuration validation in Frosting integration test project here: https://github.com/cake-build/cake/tree/develop/tests/integration/Cake.Frosting/build
> 
> It's called here:
> 
> https://github.com/cake-build/cake/blob/3531435a932dd112984f04798af694efa5326997/build.cake#L374-L384

Any example I could follow?

Wouldn't it be better if there were also unit tests?

devlead commented:
> Any example I could follow?

You can add arguments here
https://github.com/cake-build/cake/blob/3531435a932dd112984f04798af694efa5326997/build.cake#L375-L377

Environment variables can be set on the [DotNetRunSettings](https://cakebuild.net/api/Cake.Common.Tools.DotNet.Run/DotNetRunSettings/) using [EnvironmentVariables](https://cakebuild.net/api/Cake.Core.Tooling/ToolSettings/ACBC8906) property. 

Config file is here:
https://github.com/cake-build/cake/blob/3531435a932dd112984f04798af694efa5326997/tests/integration/Cake.Frosting/cake.config#L1-L2

Which then should be able to be picked up in a task here
https://github.com/cake-build/cake/tree/3531435a932dd112984f04798af694efa5326997/tests/integration/Cake.Frosting/build/Tasks


> Wouldn't it be better if there were also unit tests?

Integration test would E"E, but Unit tests won't hurt, Frosting specific tests are located here
https://github.com/cake-build/cake/blob/3531435a932dd112984f04798af694efa5326997/src/Cake.Frosting.Tests/CakeHostTests.cs

paulomorgado commented:
> > Any example I could follow?
> 
> You can add arguments here
> 
> https://github.com/cake-build/cake/blob/3531435a932dd112984f04798af694efa5326997/build.cake#L375-L377
> 
> Environment variables can be set on the [DotNetRunSettings](https://cakebuild.net/api/Cake.Common.Tools.DotNet.Run/DotNetRunSettings/) using [EnvironmentVariables](https://cakebuild.net/api/Cake.Core.Tooling/ToolSettings/ACBC8906) property.
> 
> Config file is here:
> 
> https://github.com/cake-build/cake/blob/3531435a932dd112984f04798af694efa5326997/tests/integration/Cake.Frosting/cake.config#L1-L2
> 
> Which then should be able to be picked up in a task here https://github.com/cake-build/cake/tree/3531435a932dd112984f04798af694efa5326997/tests/integration/Cake.Frosting/build/Tasks
> 
> > Wouldn't it be better if there were also unit tests?
> 
> Integration test would E"E, but Unit tests won't hurt, Frosting specific tests are located here https://github.com/cake-build/cake/blob/3531435a932dd112984f04798af694efa5326997/src/Cake.Frosting.Tests/CakeHostTests.cs

@devlead,

I couldn't figure out how to assert the results using the tasks you pointed to, but I added unit tests.

devlead committed:
Add integration tests for Frosting configuration

Add Config task to validate configuration from multiple sources:
- Environment variables (CAKE_INTEGRATIONTEST_ENVIRONMENT)
- INI file configuration (cake.config)
- Command-line arguments (--IntegrationTest_Argument)

Changes:
- Add Config task with xunit assertions to verify all config sources
- Update build script to pass test arguments and environment variables
- Add xunit.v3.assert package reference for assertions
- Fix working directory from ".." to "." in Program.cs
- Add [IntegrationTest] section to cake.config
- Use nameof() for task names instead of string literals
- Make verbosity configurable via integration-tests-verbosity argument

devlead committed:
Try address Forsting Template file in use

devlead reviewed (approved):
> I couldn't figure out how to assert the results using the tasks you pointed to, but I added unit tests.

I've added the integration tests,

LGTM now.

devlead commented:
@paulomorgado your changes have been merged, thanks for your contribution 👍

paulomorgado commented:
> @paulomorgado your changes have been merged, thanks for your contribution 👍

@devlead, thank you for your help.
